text/0000-dyn-upcasting.md

```diff
@@ -0,0 +1,289 @@
+- Feature Name: `trait_upcasting`
+- Start Date: (fill me in with today's date, YYYY-MM-DD)
+- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)
+- Rust Issue: [rust-lang/rust#65991](https://github.com/rust-lang/rust/issues/65991)
+- Design repository: [rust-lang/dyn-upcasting-coercion-initiative](https://github.com/rust-lang/dyn-upcasting-coercion-initiative)
+
+# Summary
+[summary]: #summary
+
+Enable upcasts from `dyn Trait1` to `dyn Trait2` if `Trait1` is a subtrait of `Trait2`. 
+
+This RFC does not enable `dyn (Trait1 + Trait2)` for arbitrary traits. If `Trait1` has multiple supertraits, you can upcast to any one of them, but not to all of them.
+
+This RFC has already been implemented in the nightly compiled with the feature gate `trait_upcasting`.
+
+# Motivation
+[motivation]: #motivation
+
+If you define a trait with a supertrait
+
+```rust
+trait Writer: Reader { }
+
+trait Reader { }
+```
+
+you can currently use `impl Writer` anywhere that `impl Reader` is expected:
+
+```rust
+fn writes(w: &mut impl Writer) {
+    reads(w);
+}
+
+fn reads(r: &mut impl Reader) {
+    
+}
+```
+
+but you cannot do the same with `dyn`
+
+```rust
+fn writes(w: &mut dyn Writer) {
+    reads(w); // <-- Fails to compile today
+}
+
+fn reads(r: &mut dyn Reader) {
+    
+}
+```
+
+The only upcasting coercion we permit for dyn today is to remove auto-traits; e.g., to coerce from `dyn Writer + Send` to `dyn Writer`.
+
+## Sample use case
+
+One example use case comes from the [salsa](https://github.com/salsa-rs/salsa) crate. Salsa programs have a central database but they can be broken into many modules. Each module has a trait that defines its view on the final database. So for example a parser module might define a `ParserDb` trait that contains the methods the parser needs to be present. All code in the parser module then takes a `db: &mut dyn ParserDb` parameter; `dyn` traits are used to avoid monomorphization costs.
+
+When one module uses another in Salsa, that is expressed via supertrait relationships. So if the type checker module wishes to invoke a parser, it might define its `trait TypeCheckDb: ParserDb` to have the `ParserDb` as a supertrait. The methods in the type checker then take a `db: &mut dyn TypeCheckerDb` parameter. If they wish to invoke the `ParserDb` methods, they would ideally be able to pass this `db` parameter to the parser methods and have it automatically upcast. This does not work with today's design, requiring elaborate workarounds.
```

Comment:
mismatch in the trait names

```suggestion
When one module uses another in Salsa, that is expressed via supertrait relationships. So if the type checker module wishes to invoke a parser, it might define its `trait TypeCheckerDb: ParserDb` to have the `ParserDb` as a supertrait. The methods in the type checker then take a `db: &mut dyn TypeCheckerDb` parameter. If they wish to invoke the `ParserDb` methods, they would ideally be able to pass this `db` parameter to the parser methods and have it automatically upcast. This does not work with today's design, requiring elaborate workarounds.
```